design/ceph-external-cluster.md

```diff
@@ -0,0 +1,105 @@
+   * FSID, mon endpoint(s), admin keyring
+1. The mon, mgr, and osd daemons will not be managed by the local Rook operator. These daemons must be created and managed by the external cluster.
+1. The operator will make a “best effort” to keep the list of mons updated.
+   * If the mons change in the external cluster, the list of mons must be updated in the local cluster.
+   * The operator will need to query the Ceph status periodically (perhaps every minute). If there is a change to the mons, the operator will update the local configmaps/secrets.
+   * If the local operator fails to see changes to the external mons, perhaps because it is down, the mon list could become stale. In that case, the admin will need to update the list similarly to how it was initialized when the local cluster was first created.
+
+The first bullet point above requires an extra manual configuration step by the cluster admin from what he needs in a typical Rook cluster. The other items above will be handled automatically by the Rook operator. The extra step involves exporting metadata from the external cluster and importing it to the local cluster:
+1. The admin creates a yaml file with the needed resources from the external cluster (ideally we would provide a helper script to help automate this task):
+   * Save the mon list and admin keyring
+1. Load the yaml file into the local cluster
+   * `oc create -f <config.yaml>`
+
+The CephCluster CRD will have a new property “external” to indicate whether the cluster is external. If true, the local operator will implement the described behavior.
+
+```yaml
+kind: CephCluster
+spec:
+  external: true
+```
+
+The mgr modules, including the dashboard, would be running in the external cluster. Any configuration that happens through the dashboard would depend on the orchestration modules in that external cluster.
+
+## Block Storage
+
+With the rook-ceph cluster created, the CSI driver integration will cover the Block (RWO) storage and no additional management is needed.
+
+## Filesystem (MDS)
+
+A filesystem can be created by Rook that will be specific to the local cluster. This implies that you could have multiple instances of CephFS, with one or more instances of CephFS per K8s cluster that is connecting to the external cluster.
+
+When the CephFilesystem CRD is created in the local cluster, the local Rook operator does the following:
+1. Create the data and metadata pools in the external cluster
+1. Create the filesystem in the external cluster
+1. Start the MDS in the local cluster
+1. Local clients will connect to the local MDS and the external mons and osds
+
+No code changes are necessary for this scenario. The external connections in step 1 and 2 are possible since the local operator knows the mon endpoints and has the admin keyring. For step 3, the MDS started in the local cluster won’t even know that Ceph is running externally. Again, all the MDS needs is the external mon endpoints and the keyring.
```

Comment:
It also needs access to the OSD of the external cluster.

Reply:
Does MDS need access to the OSDs? I was thinking it's just the clients.

Reply:
The MDS stores the fs metadata (inodes etc) and its journal in RADOS so it needs a connection to the OSDs.